### PR TITLE
Fix: guard around unobserveRecursively to avoid error being thrown

### PR DESCRIPTION
**Pull request by wbercx**
# What this does
When deepObserve crawls through the target, `observeRecursively` verifies whether the `thing` it finds is an observable array, -object or -map prior to observing it and going deeper. This pull request applies the same check to `unobserveRecursively`.

# Motivation
I upgraded React Native in one of my projects earlier today. One of the changes was the addition of a 64-bit release variant on the Android side. I had overridden JSC to a different version some time ago, but that ended up not being included in the 64-bit binary of my app. It used whatever version RN 0.58 ships with.

I found out that code which previously worked fine no longer did. I have a store similar to:

````js
const userStore = observable({
    models: observable.map(),
    status: 'logged-out'
})
````

I then deeply observe changes to this store.
````js
deepObserve(userStore, change => /* ... */)
````

Upon attempting to log in, I make the following change:
````js
userStore.status = 'logging-in';
````

And an error is thrown from `entrySet.get(thing)` in `unobserveRecursively`:
`A WeakMap cannot have a non-object key.`

`deepObserve` will attempt to unobserve the old value `"logged-out"` - a primitive value. According to the spec, WeakMaps only support objects as keys. It seems that most JS engines are pretty forgiving (I cannot reproduce it in V8) and just return `undefined` when you attempt to pass a string to `entrySet.get()`. 

Alas, this particular version of JSC doesn't like it _at all_ and explodes. It is impossible to remove something that cannot be added in the first place, so using the same isObservableArray|Object|Map checks should suffice.

### CODE REVIEW

**Comment by ItamarShDev** [comment]
Please add tests in 'test/deepObserve.ts'

**Comment by wbercx** [comment]
No worries. Will get on top of that tomorrow. 

Just note it won't be the most dependable test as it doesn't fail in Node in the first place.

File: src/deepObserve.ts [hunk 1]
```diff
@@ -83,15 +83,21 @@ export function deepObserve<T = any>(
...
+                    if (isRecursivelyObservable(change.object[i])) {
+                        const entry = entrySet.get(change.object[i])
+                        if (entry) entry.path = "" + i
+                    }
                 }
                 break
         }
     }
 
+    function isRecursivelyObservable(thing: any) {
```
Comment (mweststrate) [inline src/deepObserve.ts]: This function can be moved to the top level of the module, as it doesn't depend on anything in the closure

**Comment by mweststrate** [comment]
wow, great find @wbercx! Must have been tricky to figure out. Thanks for finding! Left one additional comment in addition to the tests

**Comment by wbercx** [comment]
@ItamarShDev @mweststrate Have moved `isRecursivelyObservable` out of deepObserve's closure, and added a basic test to check for throws on primitive value changes.

I added it in front of `"basic & dispose"`, as that particular test would already have thrown the error in the relevant version of JSC. Can move it down if you like.

**Comment by mweststrate** [comment]
Awesome, merged!

**Pull request merged**